src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt

```diff
@@ -0,0 +1,181 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.hints
+
+import com.intellij.codeInsight.hints.*
+import com.intellij.codeInsight.hints.presentation.InlayPresentation
+import com.intellij.codeInsight.hints.presentation.InsetPresentation
+import com.intellij.codeInsight.hints.presentation.MenuOnClickPresentation
+import com.intellij.openapi.components.service
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.DumbService
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.ui.components.CheckBox
+import com.intellij.ui.layout.panel
+import com.intellij.util.ui.JBUI
+import org.rust.lang.RsLanguage
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.ty.TyUnknown
+import org.rust.lang.core.types.type
+import javax.swing.JPanel
+
+class RsInlayTypeHintsProvider : InlayHintsProvider<RsInlayTypeHintsProvider.Settings> {
+    override val key: SettingsKey<Settings> get() = ourKey
+
+    override val name: String get() = "Type hints"
+
+    override val previewText: String?
+        get() = """
+            struct Foo<T1, T2, T3> { x: T1, y: T2, z: T3 }
+            
+            fn main() {
+                let foo = Foo { x: 1, y: "abc", z: true };
+            }
+            """.trimIndent()
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        val showForVariables = "Show for variables"
+        val showForLambdas = "Show for closures"
+        val showForIterators = "Show for iterators"
+        val showObviousTypes = "Show obvious types"
+
+        private val varField = CheckBox(showForVariables)
```

Comment:
Maybe move these settings to a separate PR?

Reply:
These settings are copied from old hints so I think we should leave them in this PR